### PR TITLE
make substitute() work for whole constraint trees

### PR DESCRIPTION
**Pull request by exaexa**
(still needs tests)

### CODE REVIEW

File: docs/src/1-metabolic-modeling.jl [hunk 1]
```diff
@@ -471,3 +471,74 @@ changes.fluxes
...
+# unused) variable index.
+#
+# First, let's create a list of all variables in the model that we can use for
+# substitution:
+
+vars = [C.variable(; idx).value for idx = 1:C.var_count(c)]
+
+# Now we use a bit of the knowledge about the model structure -- the handicaps
+# constraint single variables, so we can substitute for them directly. (If the
+# handicaps constrainted larger linear combinations of variables, we would have
```
Comment (stelmo) [inline docs/src/1-metabolic-modeling.jl:503]: "constrained"

File: docs/src/1-metabolic-modeling.jl [hunk 2]
```diff
@@ -471,3 +471,74 @@ changes.fluxes
...
+# to resort to algebra.)
```
Comment (stelmo) [inline docs/src/1-metabolic-modeling.jl:504]: maybe be more specific?

Comment (stelmo) [reply docs/src/1-metabolic-modeling.jl:504]: okay, maybe not

File: docs/src/1-metabolic-modeling.jl [hunk 3]
```diff
@@ -471,3 +471,74 @@ changes.fluxes
...
+vars[c.community.species1.handicap.value.idxs[1]] = zero(C.LinearValue)
+vars[c.community.species2.handicap.value.idxs[1]] = zero(C.LinearValue)
+
+# [`substitute`](@ref ConstraintTrees.substitute) can be used to feed these new
+# variables into all variables in the existing constraint system, which we
+# immediately follow by pruning of the variables (pruning is described closer
+# in the [functional tree processing
+# example](4-functional-tree-processing.md)).
+
+c_simplified = C.prune_variables(C.substitute(c, vars))
```
Comment (stelmo) [inline docs/src/1-metabolic-modeling.jl:518]: cool

File: docs/src/1-metabolic-modeling.jl [hunk 4]
```diff
@@ -471,3 +471,74 @@ changes.fluxes
...
+
+(C.var_count(c), C.var_count(c_simplified))
+
+@test C.var_count(c) == C.var_count(c_simplified) + 2 #src
+
+# The constraints that were substituted for are, at this point, roughly
+# equivalent to saying `0 == 0`, and most solvers will simply drop them as
+# tautologies:
+
+c_simplified.community.species1.handicap
```
Comment (stelmo) [inline docs/src/1-metabolic-modeling.jl:530]: can this be automatically removed somehow? to just get the cleaned + pruned tree

**Review by stelmo — APPROVED** [review]
good, will make diffmet easier

File: docs/src/1-metabolic-modeling.jl (hunk 4 above)
Comment (exaexa) [reply docs/src/1-metabolic-modeling.jl:530]: removal is possible (`delete!`). For automated one you'd basically need to detect tautology constraints (not complex). OTOH removing this removes the entry from the result, which might be an issue. So I wouldn't do that automatically.

File: docs/src/1-metabolic-modeling.jl (hunk 1 above)
Comment (exaexa) [reply docs/src/1-metabolic-modeling.jl:503]:
```suggestion
# handicaps constrained larger linear combinations of variables, we would have
```

**Pull request merged**